content/blog/2019-12-18-polynomial/index.md

```diff
@@ -0,0 +1,197 @@
+[TVM](https://github.com/apache/incubator-tvm). 
+
+### Spiral: FFT code generator
+Spiral is a state of the art FFT code generator. FFT can be expressed as product of sparse matrices. 
+Spiral's framework includes an DSL called SPL language that can be used to 
+described FFT's algorithm transformation.
+Currently there are around 20 break-down rules in Spiral that can be used to
+break down a transformation into a series of transformations. Using different
+rules to decomposed the matrix, we can get different implementation of the FFT
+algorithm. By trying out different combination of rules, it can find the best
+implementation on the target platform. 
+
+## Efficient Polynomial Multiplication Algorithm
+
+Inspired from Spiral, I thought that we could similarly write down the rules to 
+break down polynomial multiplication, and we can use these rules to break down polynomial
+multiplication into smaller ones. As we can see in this section, there are
+different rules we can use to recursively break down the polynomial.
+
+
+### Schoolbook Multiplication
+Suppose we want to multiply two polynomials $f(x)$ and $g(x)$, where
+  $$
+  f(x) = a_1(x) t + a_0(x)
+  $$
+  $$
+  g(x) = b_1(x) t + b_0(x)
+  $$
+
+  Then schoolbook method is to compute 
+  $$
+  f(x)g(x) = c_2(x) t^2 + c_1 t + c_0(x)
+  $$
+  where
+  $$ c_2(x) = a_1(x)b_1(x) $$
+  $$ c_1(x) = a_1(x)b_0(x) + a_0(x)b_1(x)$$
+  $$ c_0(x) = a_0(x)b_0(x) $$.
+  It can break down a large polynomial multiplications into 4 smaller polynomial
+  multiplications.
+
+### Karatsuba
+We can break the polynomial $f(x)$ and $g(x)$ to higher degree part
+and lower degree part as follows:
+
+  $$
+  f(x) = a_1(x) t + a_0(x)
+  $$
+  $$
+  g(x) = b_1(x) t + b_0(x)
+  $$
+  , where $t = x^{\frac{n}{2}}$. 
+  Then the multiplication can be calculated as follows:
+
+  $$
+  f(x)g(x) = c_2(x) t^2 + c_1 t + c_0(x)
+  $$
+
+  where 
+  $$ c_2(x) = a_1(x)b_1(x) $$
+  $$ c_0(x) = a_0(x)b_0(x) $$
+
+  $$
+  c_1(x) = (a_1(x) + a_0(x)) (b_1(x) +b_0(x)) -c_0(x)-c_2(x)
+  $$
+
+  It uses 3 multiplications while the schoolbook method would use 4.
+  The idea can be further generalized to a k-way split, and it is called Toom-k.
+
+### Toom-k
+
+Lets take Toom-3 as an example. First we divide the polynomial $f(x)$ and $g(x)$ into 3 parts.
+
+$$
+f(x) = a_2(x) t^2 + a_1(x) t + a_0(x)
+$$
+
+$$
+g(x) = b_2(x) t^2 + b_1(x) t + b_0(x)
+$$
+
+where $t = x^{\frac{n}{3}}$. 
+Similarly to Karatsuba method, in this case, the number of
+multiplication can be reduced to 5, instead of 9 in the case of schoolbook
+method and asymptotically better than both Karatsuba method and schoolbook
+method. This method can be generalized to $k$-way split.
+
+### Unable to decouple the algorithm from the order of computation
+However, I am unable to find a way to decouple the above algorithm from the
+order of computation while still be easy to configure and be useful. As a
+result, it makes this project unsuccessful.
```

Comment:
I'm not entirely sure what this means… what would it look like to "decouple" the "order of computation"? I'm not exactly sure what different orders of computation would offer, or how they would be specified. This notion seems to be new in this section, so I'm not really sure how to interpret it.